### PR TITLE
Translate `_d_newThrowable` hook to template

### PR DESCRIPTION
**Pull request by teodutu**
This PR lowers expressions such as `throw new E(args)` to:
```d
throw _d_newThrowable!E(args);
``` 
when compiling with `-dip1008`.

The DRuntime PRs which implement the template `_d_newThrowable` are:
- https://github.com/dlang/druntime/pull/3661
- https://github.com/dlang/druntime/pull/3692

### CODE REVIEW

**Comment by thewilsonator** [comment]
dependant PR merged, please rebase this.

**Review by thewilsonator — COMMENTED** [review]
is this able to be tested in this PR?

**Comment by teodutu** [comment]
@thewilsonator Yes. This lowering is enabled by `-dip1008`. The following tests use this parameter:
- test/compilable/test20318.d
- test/fail_compilation/test20023.d
- test/runnable/test19317.d
- test/compilable/testdip1008.d

**Comment by thewilsonator** [comment]
There are a few build kite failures, could you take a look. cc @atilaneves this fails on your projects

**Comment by atilaneves** [comment]
> There are a few build kite failures, could you take a look. cc @atilaneves this fails on your projects

Yes - but I don't know why it'd fall on me to go and see why. I downloaded the branch from the PR anyway and tried it on unit-threaded, and it seems this PR ends up changing the purity of `std.format.spec.FormatSpec!char.FormatSpec.toString!(Appender!string).toString` to impure.

The buildkite test projects are in CI exactly to try and catch this type of breakage.

I'm more than happy to help and answer questions because of familiarity (or lack thereof) with the codebase.

**Comment by RazvanN7** [comment]
Rebase this and let's get it in.

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -4866,6 +4866,66 @@ public:
...
+                        }
+                        else
+                            m = v.type.defaultInitLiteral(e.loc);
+                        if (exceptionOrCant(m))
+                            return;
+                        (*elems)[fieldsSoFar + i] = copyLiteral(m).copy();
+                    }
+                }
+
+                // Hack: store a ClassDeclaration instead of a StructDeclaration.
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d:4914]: This seems to be storing a StructDeclaration instead of a ClassDeclaration, not the other way around.

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -4866,6 +4866,66 @@ public:
...
+                auto se = ctfeEmplaceExp!StructLiteralExp(e.loc, cast(StructDeclaration) cd, elems, e.type);
+                se.origin = se;
+                se.ownedByCtfe = OwnedBy.ctfe;
+                Expression eref = ctfeEmplaceExp!ClassReferenceExp(e.loc, se, e.type);
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d:4918]: Do you mind explaining why is this necessary?

File: src/dmd/dinterpret.d [hunk 3]
```diff
@@ -4866,6 +4866,66 @@ public:
...
+
+                UnionExp ue = void;
+                auto err = interpretFunction(&ue, cast(FuncDeclaration) cd.ctor, istate, e.arguments, eref);
+                if (exceptionOrCant(err))
+                    return;
+
+                eref.loc = e.loc;
+                result = eref;
+                return;
+            }
```
Comment (MoonlightSentinel) [inline src/dmd/dinterpret.d:4928]: This block seems to be copied from the class branch of `NewExp` - refactor to that code into a function instead of duplicating it here.

File: src/dmd/e2ir.d [hunk 4]
```diff
@@ -1096,8 +1096,11 @@ extern (C++) class ToElemVisitor : Visitor
             }
             else
             {
+                assert(!(global.params.ehnogc && ne.thrownew),
+                    "This should have been rewritten to `_d_newThrowable` in the semantic phase.");
```
Comment (MoonlightSentinel) [inline src/dmd/e2ir.d:1100]: This assert will trigger for COM / C++ classes which are not rewritten in `expressionsem.d`?

**Review by MoonlightSentinel — DISMISSED** [review]

File: src/dmd/dinterpret.d (hunk 2 above)
Comment (teodutu) [reply src/dmd/dinterpret.d:4918]: `eref` needs to be a reference to `this` of the constructed class, in order to call `interpretFunction` below. This is why it needs to be a `ClassReferenceExp`. The ctor of `ClassReferenceExp` receives a `StructLiteralExp`. Hence, the need for the "hack".

File: src/dmd/dinterpret.d (hunk 3 above)
Comment (teodutu) [reply src/dmd/dinterpret.d:4928]: Wouldn't it be shorter and more expressive, albeit a bit less efficient, if I just create a `NewExp` and interpret it directly?

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (teodutu) [reply src/dmd/dinterpret.d:4914]: You're right. Fixed.

File: src/dmd/e2ir.d (hunk 4 above)
Comment (teodutu) [reply src/dmd/e2ir.d:1100]: You're right. I changed this `assert` to the same condition as in expressionsem.d: https://github.com/dlang/dmd/blob/bf6d206751c4a5bcf20ab4c19e4a7ff283a318dc/src/dmd/expressionsem.d#L3727-L3728.

**Comment by RazvanN7** [comment]
It seems that this is still failing a couple of projects on buildkite

**Comment by kinke** [comment]
Looks like the undefined symbol errors are caused by the same underlying issue as using `-checkaction=context`. E.g., failing `unit-threaded` uses `-preview=dip1008`: https://github.com/atilaneves/unit-threaded/blob/e9e612b5126b3c99c8239f2e09b65c9303e53d94/dub.json#L51

The compiler assumes druntime and Phobos are compiled with the same switch, and so most likely skips codegen for some templates encountered in druntime/Phobos imports. So once again a template culling problem when it comes to linking libraries compiled with different switches.

**Comment by RazvanN7** [comment]
@kinke : @teodutu compiled druntime locally with dip1008 enabled and then compiled unit-threaded with this version however the linker error persisted. One thing we could try is to compile both druntime and phobos with dip1008 and see if that works, however, phobos uses exceptions a lot and underlying dip1008 issues might surface (I remember looking at some dip1008 issue a few years ago and my thought back then is that it works for simple cases, but there are more complex uses where it's gonna fail; IIRC it was something regarding scope exceptions that could/should have escaped their scope). But if that doesn't work, what else can we do?

**Comment by kinke** [comment]
I don't know all the DIP1008 details, but to me it seemed to be just a way to use exceptions with `-betterC`, without GC and without druntime/Phobos. So compiling druntime/Phobos with DIP1008 doesn't seem to make much sense to me (not linked with `-betterC` anyway).

Besides fixing `-allinst`, another option might be to use `pragma(inline, true)` for `_d_newThrowable`. Not sure how that is implemented by DMD, but with LDC, that should work in the sense of truly inlining it for each all call, in each compiled object file.

**Comment by teodutu** [comment]
@kinke Using `pragma(inline, true)` didn't solve the linking error. I guess the symbol that causes it stays the same regardless of inlining.

**Comment by kinke** [comment]
Oh sorry, completely forgot that the problem is the nested template, `forward`. Hmm, that's called from user code too, so force-inlining that for debug builds would probably be too much. So I don't really see an alternative to `-allinst`, or some lighter, ugly special-casing of `_d_newThrowable` templates during `needsCodegen()`, similar to #13599.

**Comment by thewilsonator** [comment]
Please rebase.

File: src/dmd/e2ir.d (hunk 4 above)
Comment (RazvanN7) [reply src/dmd/e2ir.d:1076]: It doesn't seem that the `!isCOMclass() && !cd.isCPPclass()` condition did not make it.

**Review by RazvanN7 — APPROVED** [review]
Don't forget to delete the old hook in druntime!

**Comment by nordlow** [comment]
How does this improve things for the developer?

**Comment by teodutu** [comment]
@nordlow It's pretty transparent from the perspective of the developer. The PR makes the code in the compiler easier to maintain and contributes to https://github.com/dlang/projects/issues/25. The issue only mentions arrays, but there are other cases where `TypeInfo` is used, like this one.
